Add basic StatusService

joshdover opened the pull request:
## Summary

Related to #41983 
Follows the spec outlined in RFC-0010.

This implements the basics of the StatusService + exposes the core status of the Elasticsearch and SavedObject services. These values are not currently being read by anything in the system yet, but will be in a future change.

### Checklist

Delete any items that are not applicable to this PR.

- [x] [Unit or functional tests](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#cross-browser-compatibility) were updated or added to match the most common scenarios
### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://github.com/elastic/kibana/blob/master/CONTRIBUTING.md#release-notes-process)

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/)
* Commit: 8cc03f30dd7ee906addbb88d593270fd9044b4b7

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/","number":"33634","commit":"8cc03f30dd7ee906addbb88d593270fd9044b4b7"}]}
PIPELINE-->

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/)
* Commit: 8d567b7234e85ec73e67684e471e28329beda2fe

### History
* :broken_heart: [Build #33634](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/) failed 8cc03f30dd7ee906addbb88d593270fd9044b4b7

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/","number":"33634","commit":"8cc03f30dd7ee906addbb88d593270fd9044b4b7"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/","number":"34746","commit":"8d567b7234e85ec73e67684e471e28329beda2fe"}]}
PIPELINE-->

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34762/)
* Commit: ed6c0801fe328c20e584b6d7226005dce368f8a5

### History
* :broken_heart: [Build #34746](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/) failed 8d567b7234e85ec73e67684e471e28329beda2fe
* :broken_heart: [Build #33634](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/) failed 8cc03f30dd7ee906addbb88d593270fd9044b4b7

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/","number":"33634","commit":"8cc03f30dd7ee906addbb88d593270fd9044b4b7"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/","number":"34746","commit":"8d567b7234e85ec73e67684e471e28329beda2fe"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34762/","number":"34762","commit":"ed6c0801fe328c20e584b6d7226005dce368f8a5"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-platform (Team:Platform)

pgayvallet commented on src/core/server/status/types.ts:
```diff
@@ -0,0 +1,126 @@
+      summary: string; // required when level !== available
+      detail?: string;
+      documentationUrl?: string;
+      meta?: Meta;
```
NIT: these three last properties are the same in both cases. They could be factorized, even if I'm not sure this would increase readability on this already complex type.

joshdover replied:
Resolved by reducing to a single interface with `level` and `summary` always required.

pgayvallet commented on src/core/server/status/types.ts:
```diff
@@ -0,0 +1,126 @@
+  // TODO
+  // http: ServiceStatus;
+  // uiSettings: ServiceStatus;
+  // metrics: ServiceStatus;
```
Are these going to be in this PR or in a follow-up?

joshdover replied:
In follow up. I will remove for now.

pgayvallet commented on src/core/server/status/types.ts:
```diff
@@ -0,0 +1,126 @@
+  /**
+   * Current status for all Core services.
+   */
+  core$: Observable<CoreStatus>;
```
NIT: extends `StatusServiceSetup` ?

pgayvallet commented on src/core/server/status/types.ts:
```diff
@@ -0,0 +1,126 @@
+  /**
+   * Overall system status used for HTTP API
+   */
+  overall$: Observable<ServiceStatus>;
```
Plugins have no access to that?

joshdover replied:
No, this will only be used by the Status API. I don't want plugins to be able to get into some circular loop by allowing them to read the overall status which leak status info from plugins they do not explicitly depend on. If there's a valid use case though, let me know.

rudolf commented on src/core/server/saved_objects/status.ts:
```diff
@@ -0,0 +1,75 @@
+  return combineLatest(elasticsearchStatus$, migratorStatus$).pipe(
+    map(([esStatus, migratorStatus]) => {
+      if (esStatus.level >= ServiceStatusLevel.unavailable) {
+        return {
```
I agree that the observables that we share from the status service within core and with plugins needs to be hot and always have the latest value for new subscribers. The status page will probably wait for an emission from each observable to render the page. 

When we expose this to plugins the status service's method `set(status$: Observable<ServiceStatus>): void;` can turn plugin `status$` observables into a hot observable before further distributing it, but that's future work.

joshdover replied:
Agreed, this should be at the StatusService level. I will go ahead and add a replay to the `overall$` API which will be exposed to the status page when it is migrated.

restrry commented on src/core/server/status/status_service.ts:
```diff
@@ -0,0 +1,74 @@
+  public stop() {}
+
+  private setupCoreStatus(coreStatuses: SetupDeps['coreStatuses']): Observable<CoreStatus> {
+    return combineLatest(coreStatuses.elasticsearch$, coreStatuses.savedObjects$).pipe(
```
```suggestion
    return combineLatest([coreStatuses.elasticsearch$, coreStatuses.savedObjects$]).pipe(
```

restrry commented on src/core/server/saved_objects/status.ts:
```diff
@@ -0,0 +1,75 @@
+      if (esStatus.level >= ServiceStatusLevel.unavailable) {
+        return {
+          level: ServiceStatusLevel.unavailable,
+          summary: `SavedObjects are not available without a healthy Elasticearch connection`,
```
```suggestion
          summary: `SavedObjects service is not available without a healthy Elasticsearch connection`,
```

restrry commented on src/core/server/saved_objects/status.ts:
```diff
@@ -0,0 +1,75 @@
+      } else if (esStatus.level === ServiceStatusLevel.degraded) {
+        return {
+          level: esStatus.level,
+          summary: esStatus.summary,
```
should provide a summary of the own status instead of elasticsearch?

restrry commented on src/core/server/elasticsearch/elasticsearch_service.ts:
```diff
@@ -186,6 +187,7 @@ export class ElasticsearchService
       adminClient: this.adminClient,
       dataClient,
       createClient: this.createClient,
+      status$: calculateStatus$(esNodesCompatibility$),
```
now we can stop exporting  `esNodesCompatibility$` ? https://github.com/elastic/kibana/blob/960d8a3c305f52307b298746d660777a58e71f00/src/core/server/saved_objects/saved_objects_service.ts#L400-L405

joshdover replied:
It's still being used by the legacy ES plugin. I was going to switch that over in a separate PR.

restrry commented on src/core/server/saved_objects/status.ts:
```diff
@@ -0,0 +1,75 @@
+    map(migrationResult => {
+      const skipped = migrationResult.filter(({ status }) => status === 'skipped').length;
+      const patched = migrationResult.filter(({ status }) => status === 'patched').length;
+      const migrated = migrationResult.filter(({ status }) => status === 'migrated').length;
```
Is there a way to make sure we handle **all possible** statuses?

restrry commented on src/core/server/saved_objects/status.ts:
```diff
@@ -0,0 +1,75 @@
+        return migratorStatus;
+      }
+    }),
+    distinctUntilChanged(isDeepStrictEqual)
```
`distinctUntilChanged` can be moved to status service. It's easy to forget to add it to the plugin code.

restrry commented on src/core/server/status/status_service.mock.ts:
```diff
@@ -0,0 +1,71 @@
+const createSetupContractMock = () => {
+  const setupContract: jest.Mocked<StatusServiceSetup> = {
+    core$: mockCoreStatus,
+  };
```
or provide a mockable getter interface? the current behavior is okay for me, as it's compatible with the licensing service. if we adopt another solution, I'd unify this approach across our codebase

joshdover replied:
I like licensing approach of using a BehaviorSubject which allows the test consumer to push new values into if it wishes (though with some typecasting). I'll go with this for now.

joshdover commented on src/core/server/status/types.ts:
```diff
@@ -0,0 +1,126 @@
+    }
+  | {
+      level: ServiceStatusLevel;
+      summary: string; // required when level !== available
```
I'm good with this. TypeScript has some issues with inference on this type as well. I was getting very hard-to-understand type errors when working with the status checks in SO and ES services. Having a single type will make this simpler.

joshdover commented on src/core/server/elasticsearch/status.test.ts:
```diff
@@ -0,0 +1,218 @@
+    expect(statusUpdates.map(({ level, summary }) => ({ level, summary }))).toMatchInlineSnapshot(`
+      Array [
+        Object {
+          "level": 2,
```
Yeah I think the sorting property that I liked from a regular integer enum can be expressed as a utility. String readability is probably more important. 👍

joshdover replied:
Actually on second thought, I think we're more often concerned with comparisons than we are with the string output (which is really only needed in the API outputs).

If we move to string enums, we would probably want to include a utility like this:
```ts
export const statusComparator = {
  gt: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) === 1,
  gte: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) >= 0,
  lt: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) === -1,
  lte: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) <= -1,
  eq: (a: ServiceStatusLevel, b: ServiceStatusLevel): boolean => compareFn(a, b) === 0,
  compareFn,
};
```

However, I think this will be far less discoverable. Since doing comparisons like `esStatus >= unavailable` may be very common, I don't think this is an improvement. WDYT?

restrry replied:
> However, I think this will be far less discoverable. Since doing comparisons like esStatus >= unavailable may be very common, I don't think this is an improvement. WDYT?

Optional: We can define each level as an object or a class instance with a `valueOf` method to make them comparable.
```js
const available = {
  toString() {
    return 'available';
  },
  valueOf() {
    return 0;
  },
};

const degraded = {
  toString() {
    return 'degraded';
  },
  valueOf() {
    return 1;
  },
};
const levels = {
  available,
  degraded,
};

console.log(levels.available === levels.available);
console.log(levels.available > levels.degraded);
console.log(levels.available === levels.degraded);
console.log(levels.available < levels.degraded);
```

pgayvallet replied:
```ts
const available = {
  toString() {
    return 'available';
  },
  valueOf() {
    return 0;
  },
};
```

Would love to have java `enum` class implementation in TS. that would really solves this kind of issues in a graceful way...

joshdover replied:
Good idea, switched to this pattern. It doesn't work for array sorting, but I'm less concerned about that being a common use case.

joshdover commented on src/core/server/saved_objects/status.ts:
```diff
@@ -0,0 +1,75 @@
+        },
+      };
+    }),
+    startWith({
```
Added two different statuses, one that explains that SO migrations are waiting and one that explains when they are currently running.

joshdover commented on src/core/server/server.ts:
```diff
@@ -144,15 +147,23 @@ export class Server {
+        elasticsearch$: elasticsearchServiceSetup.status$,
+        savedObjects$: savedObjectsSetup.status$,
+      },
+    });
```
Yep, good call.

joshdover commented on src/core/server/status/types.ts:
```diff
@@ -0,0 +1,126 @@
+  // uiSettings: ServiceStatus;
+  // metrics: ServiceStatus;
+
+  // Allows CoreStatus to be used as a Record in utility functions/
```
Changed to use typecasting in service status.

joshdover commented:
Updated addressing all comments. The only thing that I think needs discussion is whether we should use integer enums or a string enum.

I prefer the integer enum since it allows for easy comparison between status levels, which is very convenient. As [I outlined here](https://github.com/elastic/kibana/pull/60335#discussion_r400474079) we could provide utilities to make this easier with string enums, however I worry that it won't be as easily discoverable.

kibanamachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/37319/)
* Commit: 1fef04aef8a6fb7d8f9af3537e284fb1b400757e
* [Pipeline Steps](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/37319/flowGraphTable) (look for red circles / failed steps)
* [Interpreting CI Failures](https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html)

### Failed CI Steps
* [Execute kibana-intake](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/37319/execution/node/153/log)

### History
* :green_heart: [Build #34762](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34762/) succeeded ed6c0801fe328c20e584b6d7226005dce368f8a5
* :broken_heart: [Build #34746](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/) failed 8d567b7234e85ec73e67684e471e28329beda2fe
* :broken_heart: [Build #33634](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/) failed 8cc03f30dd7ee906addbb88d593270fd9044b4b7

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/","number":"33634","commit":"8cc03f30dd7ee906addbb88d593270fd9044b4b7"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/","number":"34746","commit":"8d567b7234e85ec73e67684e471e28329beda2fe"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34762/","number":"34762","commit":"ed6c0801fe328c20e584b6d7226005dce368f8a5"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/37319/","number":"37319","commit":"1fef04aef8a6fb7d8f9af3537e284fb1b400757e"}]}
PIPELINE-->

restrry commented on src/core/server/status/status_service.ts:
```diff
@@ -0,0 +1,79 @@
+    const core$ = this.setupCoreStatus(core);
+    const overall$: Observable<ServiceStatus> = core$.pipe(
+      map(coreStatus => {
+        const summary = getSummaryStatus(coreStatus as any);
```
why `as any`?

joshdover replied:
Without an index signature, this has be cast to `Record<string, ServiceStatus>` in order to be compatible with the getSummaryStatus API.

To make this better, I've changed this API to accept a `Array<[string, ServiceStatus]>` which can easily be generated with `Object.entries`

pgayvallet commented on src/core/server/saved_objects/migrations/kibana/kibana_migrator.mock.ts:
```diff
@@ -47,6 +48,20 @@ const createMigrator = (
+          result: [
+            {
+              status: 'migrated',
+              destIndex: '.test-kibnaa_2',
```
NIT: typo in `kibnaa`

pgayvallet commented on src/core/server/saved_objects/migrations/kibana/kibana_migrator.ts:
```diff
@@ -109,12 +124,19 @@ export class KibanaMigrator {
     Array<{ status: string }>
   > {
     if (this.migrationResult === undefined || rerun) {
+      this.status$.next({ status: 'running' });
       this.migrationResult = this.runMigrationsInternal();
     }
```
I'm mixed about the fact to re-emit `running` status in case of `runMigrations({ rerun: true})`. OTOH, this should only be used for FTR tests, so I guess that's alright.

rudolf commented on src/core/server/status/status_service.ts:
```diff
@@ -44,13 +44,16 @@ export class StatusService implements CoreService<InternalStatusServiceSetup> {
+        this.logger.debug(`Recalculated overall status`, { status: summary });
+        return summary;
+      }),
+      distinctUntilChanged(isDeepStrictEqual),
+      shareReplay(1)
```
We do another distinctUntilChanged and shareReplay in `setupCoreStatus` so we can remove this one.

joshdover replied:
Maybe the `shareReplay` could be removed but I think we could still produce duplicate statuses since the summary rollup may or may not be the same value.

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39677/)
* Commit: 547162847b41f2756c3221c474d79a9d72b04880

### History
* :broken_heart: [Build #37319](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/37319/) failed 1fef04aef8a6fb7d8f9af3537e284fb1b400757e
* :green_heart: [Build #34762](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34762/) succeeded ed6c0801fe328c20e584b6d7226005dce368f8a5
* :broken_heart: [Build #34746](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/) failed 8d567b7234e85ec73e67684e471e28329beda2fe
* :broken_heart: [Build #33634](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/) failed 8cc03f30dd7ee906addbb88d593270fd9044b4b7

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/33634/","number":"33634","commit":"8cc03f30dd7ee906addbb88d593270fd9044b4b7"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34746/","number":"34746","commit":"8d567b7234e85ec73e67684e471e28329beda2fe"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/34762/","number":"34762","commit":"ed6c0801fe328c20e584b6d7226005dce368f8a5"},{"status":"FAILURE","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/37319/","number":"37319","commit":"1fef04aef8a6fb7d8f9af3537e284fb1b400757e"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/39677/","number":"39677","commit":"547162847b41f2756c3221c474d79a9d72b04880"}]}
PIPELINE-->
